ext/mysqlnd/mysqlnd_commands.c

```diff
@@ -642,13 +643,12 @@ MYSQLND_METHOD(mysqlnd_command, handshake)(MYSQLND_CONN_DATA * const conn, const
 	conn->protocol_version	= greet_packet.protocol_version;
 	conn->server_version	= mnd_pestrdup(greet_packet.server_version, conn->persistent);
 
-	conn->greet_charset = mysqlnd_find_charset_nr(greet_packet.charset_no);
-	if (!conn->greet_charset) {
-		char * msg;
-		mnd_sprintf(&msg, 0, "Server sent charset (%d) unknown to the client. Please, report to the developers", greet_packet.charset_no);
-		SET_CLIENT_ERROR(conn->error_info, CR_NOT_IMPLEMENTED, UNKNOWN_SQLSTATE, msg);
-		mnd_sprintf_free(msg);
-		goto err;
+	const MYSQLND_CHARSET *read_charset = mysqlnd_find_charset_nr(greet_packet.charset_no);
+	if (!read_charset) {
```

Comment:
If you look closely at `mysqlnd_charset.c`, there are some missing numbers in the character set ID. (e.g. 76, 100-127)

So, checking for `0` is not enough here. We need to check whether the corresponding character set exists.

Reply:
Yes, Saki. That's why I feel that the way the current code works is correct. It raises an error whenever an unexpected charset is provided. 0 doesn't have any special meaning here. 

What you are suggesting would probably be ok, but not ideal. `utf8` and `utf8mb4` should be a viable fallback value. The advantage is that we open the connection despite receiving an invalid charset value from the server.  The disadvantage is that we silently switch to a different charset than a user might have expected. A properly designed PHP application will set the desired charset anyway. 

So I would suggest using the two constant values as a fallback if we receive an unknown value from the server. Then we can think if we can drop this feature from mysqlnd in a major release. Maybe 9.0 would be good to do this.

Reply:
Thank you for confirmation.

> So I would suggest using the two constant values as a fallback if we receive an unknown value from the server.

If I didn't make a mistake, it should already work as a fallback like you mention. Two constants are used when `mysqlnd_find_charset_nr()` cannot find a character set. If it can be found, it uses the charset it finds in L651.

Reply:
> Then we can think if we can drop this feature from mysqlnd in a major release. Maybe 9.0 would be good to do this.

Agree. Initially I was thinking of using 8.4, but 9.x might be ideal.